src/ecc.rs

```diff
@@ -0,0 +1,425 @@
+        self.exponent1.zeroize();
+        self.exponent2.zeroize();
+        self.coefficient.zeroize();
+    }
+}
+
+impl DERMinEccPkey {
+    pub fn new(
+        modulus: &Vec<u8>,
+        public_exponent: &Vec<u8>,
+        private_exponent: &Vec<u8>,
+        prime1: &Vec<u8>,
+        prime2: &Vec<u8>,
+        exponent1: &Vec<u8>,
+        exponent2: &Vec<u8>,
+        coefficient: &Vec<u8>,
+    ) -> DERMinRsaPkey {
+        DERMinRsaPkey {
+            modulus: val_to_dermin!(modulus),
+            public_exponent: val_to_dermin!(public_exponent),
+            private_exponent: val_to_dermin!(private_exponent),
+            prime1: val_to_dermin!(prime1),
+            prime2: val_to_dermin!(prime2),
+            exponent1: val_to_dermin!(exponent1),
+            exponent2: val_to_dermin!(exponent2),
+            coefficient: val_to_dermin!(coefficient),
+        }
+    }
+
+    pub fn to_asn1(&self) -> KResult<Vec<u8>> {
+        let pkey = RSAPrivateKey {
+            version: 0,
+            modulus: biguint_or_err!(self.modulus),
+            public_exponent: biguint_or_err!(self.public_exponent),
+            private_exponent: biguint_or_err!(self.private_exponent),
+            prime1: biguint_or_err!(self.prime1),
+            prime2: biguint_or_err!(self.prime2),
+            exponent1: biguint_or_err!(self.exponent1),
+            exponent2: biguint_or_err!(self.exponent2),
+            coefficient: biguint_or_err!(self.coefficient),
+            other_prime_infos: None,
+        };
+        match asn1::write_single(&pkey) {
+            Ok(x) => Ok(x),
+            Err(_) => err_rv!(CKR_GENERAL_ERROR),
+        }
+    }
+}
+*/
+
+impl ObjectFactory for ECCPrivFactory {
+    fn create(&self, template: &[CK_ATTRIBUTE]) -> KResult<Object> {
+        let mut obj = self.default_object_create(template)?;
+
+        ecc_import(&mut obj)?;
+
+        Ok(obj)
+    }
+
+    fn get_attributes(&self) -> &Vec<ObjectAttr> {
+        &self.attributes
+    }
+
+    /*
+    fn export_for_wrapping(&self, key: &Object) -> KResult<Vec<u8>> {
+        PrivKeyFactory::export_for_wrapping(self, key)
+    }
+
+    fn import_from_wrapped(
+        &self,
+        data: Vec<u8>,
+        template: &[CK_ATTRIBUTE],
+    ) -> KResult<Object> {
+        PrivKeyFactory::import_from_wrapped(self, data, template)
+    }*/
+}
+
+impl CommonKeyFactory for ECCPrivFactory {}
+
+impl PrivKeyFactory for ECCPrivFactory {
+    /*
+    fn export_for_wrapping(&self, key: &Object) -> KResult<Vec<u8>> {
+    }
+
+    fn import_from_wrapped(
+        &self,
+        data: Vec<u8>,
+        template: &[CK_ATTRIBUTE],
+    ) -> KResult<Object> {
+    }
+    */
+}
+
+static PUBLIC_KEY_FACTORY: Lazy<Box<dyn ObjectFactory>> =
+    Lazy::new(|| Box::new(ECCPubFactory::new()));
+
+static PRIVATE_KEY_FACTORY: Lazy<Box<dyn ObjectFactory>> =
+    Lazy::new(|| Box::new(ECCPrivFactory::new()));
+
+fn check_key_object(key: &Object, public: bool, op: CK_ULONG) -> KResult<()> {
```

Comment:
should I make this stuff available in a util module so you can simply reuse it here instead of copy/paste?

Reply:
Yes, if we add a key type argument, I think it should do the generic enough for all key types we will need.

Reply:
I pushed a patch that provides an Object function for that.

Reply:
Rebased and resolved.